[Cache] handle redis cluster creation by dsn

alekitto opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| License       | MIT
| Doc PR        | TBD

This PR adds support for `\RedisCluster` connection creation via DSN passing `redis+cluster` as protocol (ex: `redis+cluster://redis-server-1:6379,redis-server-2:6379`).

This allow to have the same configuration for different environments and change only the env var containing the connection string.

nicolas-grekas commented on src/Symfony/Component/Cache/Traits/RedisTrait.php:
```diff
@@ -233,9 +258,33 @@ protected function doClear($namespace)
+        } elseif ($this->redis instanceof \RedisCluster || $this->redis instanceof RedisClusterProxy) {
+            $keys = array();
+
+            foreach ($this->redis->_masters() as $nodeParams) {
```
Coincidently, I did also fix doClear in #28269, with a slightly different implementation, see
https://github.com/symfony/symfony/pull/28269/files#diff-a87375907f78e0dcbee340d314e3d052L236
For this reason, this should be reverted from this PR.

alekitto replied:
Ok, i've reverted it.

nicolas-grekas commented on src/Symfony/Component/HttpFoundation/Session/Storage/Handler/RedisSessionHandler.php:
```diff
@@ -45,7 +46,8 @@ public function __construct($redis, array $options = array())
             !$redis instanceof \Predis\Client &&
-            !$redis instanceof RedisProxy
+            !$redis instanceof RedisProxy &&
+            !$redis instanceof RedisClusterProxy
```
the Lock component also needs a similar patch

alekitto replied:
👍 Done

nicolas-grekas reviewed (commented):
Nice, thanks!
Here are some 1st review comments.
I'm not sold on the "redis+cluster" scheme. The reason is that when using Predis, there is another way to configure a cluster proposed in #28175 
It would be awesome to be able to have a similar way of declaring a cluster when using either RedisCluster from the extension or from Predis.
Maybe let's merge #28175 as is, rebase this PR on top and refactor it to handle a common URL syntax?

alekitto commented:
I've rebased this PR on master and fixed some conflicts.
For the url syntax question, what about `cluster=1` or `cluster=true` option in query string instead of `redis+cluster` scheme?

jralph commented:
Would `cluster=` need to be a string? With redis you generally have 2 cluster options, handle the cluster from the client side, or let redis handle the cluster. So maybe a `cluster=client` and a `cluster=server` options. If cluster is `server`, then RedisCluser or Predis can be used, if it is `client`, then I believe only Predis would work.

nicolas-grekas commented:
Yep, there are two modes to select from, both legit choices; `cluster=client` also works with the extension: that's what the `RedisArray` class is for.

nicolas-grekas commented:
Something like that?

`redis://localhost:1234?server_version=2.8&cluster=array&server[host1:port1]&server[host2:port2]=weight2`

would work also for memcached:
`memcached://localhost:1234?server_version=2.8&cluster=array&server[host1:port1]&server[host2:port2]=weight2`

alekitto committed:
add support for redis cluster

alekitto commented:
I’ve implemented the `cluster=server` part, while I think that the array option should be done in a separate PR since other options could be added/discussed (the “previous” array ring option for example)

jralph commented:
With regards to using `predis` for clustering, there is a downside I only recently found. Currently, the connection factory predis provides doesn't support creating a redis cluster connection.

For clustering to work with predis, the connection dsn's must be passed in as an array instead of using the factory, or the `cluster=redis` option is just plain ignored.

alekitto committed:
cluster support for predis dsn

jralph commented on src/Symfony/Component/Cache/Traits/RedisTrait.php:
```diff
@@ -182,10 +182,25 @@ public static function createConnection($dsn, array $options = array())
+
+                // Predis cluster only supports an array of hosts as first argument, otherwise
+                // options array is ignored.
+                $redis = new $class($host, array('cluster' => 'redis'));
```
I believe `cluster=redis` should be based on the `cluster` dsn parameter. If the dsn parameter is `server`, then here it should be redis`, if the dsn parameter is `client`, then here it should be `predis`.

alekitto replied:
Like I said in a previous comment, the client part has been not implemented as could involve some other options to be handled, so I think it should be done and discussed in another PR

nicolas-grekas commented:
Closing in favor of #28713
Please review/test this PR and see if this works for you.
It'd be awesome if you could contribute some doc for it :)
